### PR TITLE
Add repair share functionality

### LINKED ISSUE #41: feature: Share redistribution - allow t signers to add new members and repair a user share
## Motivation

We want to be able to:
- Repair a user share (recompute it if it's lost)
- Add new signers - Just "recover" the share of the new user (e.g. using its identifier)

## Design

- Implement the Enrolment scheme from https://eprint.iacr.org/2017/1155.pdf
(or the Reduced Enrolment scheme; evaluate which one is better)
- Like the rest of the FROST implementation, communication is up to the caller, so we just want to provide functions to repair and add signers
- Write a test, e.g. simulating the communication using a HashMap

Preliminary design of functions to be implemented:

```

- helpers: Set of identifiers (signers that will help recover)
- requester: The identifier of the signer that lost their share


# For every single helper i in helpers:

# i: the identifier of the signer helping
# helpers: as above
# share_i: i's secret share
# zeta_i: Lagrange coefficient (?)
# - Note: may be able to be computed inside the function, check
# Output: i_deltas: random values that sum up to zeta_i * share _i
compute_random_values(i, helpers, share_i, zeta_i) -> deltas_i

# Communication round:
# Helper i sends deltas_i[j] to helper j

# j: the identifier of the signer helping
# helpers: as above
# deltas_j: values received by j in the communication round
# Output: sigma_j
compute_sum_of_random_values(j, helpers, deltas_j) -> sigma_j

# Communication round
# Helper j sends sigma_j to signer r

# sigmas: all sigma_j received from each helper j
# Output: share_r: r's secret share
recover_share(sigmas) -> share_r
```

### Specifications

See https://github.com/chelseakomlo/talks/blob/master/2019-combinatorial-schemes/A_Survey_and_Refinement_of_Repairable_Threshold_Schemes.pdf

### PR DESCRIPTION
**Pull request by natalieesk**
This closes #41

### CODE REVIEW

File: frost-core/src/frost/keys/repairable.rs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+/// For every single helper i in helpers: generate random values (deltas) for each helper
+
+/// # helpers: identifiers for signers involved in recovering share for participant
+/// # share_i: i's secret share
+/// # participant: signer who
+/// # helper_i: the identifier of the signer helping
+/// # Output: i_deltas: random values that sum up to zeta_i * share_i
+
+pub fn generate_deltas_to_repair_share<C: Ciphersuite, R: RngCore + CryptoRng>(
```
[the remark this answers is not in the document]
Comment (natalieesk) [reply frost-core/src/frost/keys/repairable.rs:17]: Yeh I agree. I went back and forth on the naming but I can try again.

File: frost-core/src/frost/keys/repairable.rs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        sum_i_deltas = sum_i_deltas + *v;
+    }
+
+    out.insert(helpers[helpers.len() - 1], lhs - sum_i_deltas);
+
+    out
+}
+
+/// TODO - Nat: This should be moved and made private
+pub fn compute_lagrange_coefficient<C: Ciphersuite>(
```
[the remark this answers is not in the document]
Comment (natalieesk) [reply frost-core/src/frost/keys/repairable.rs:59]: I think that's a separate PR. I can make that.

File: frost-core/src/frost/keys/repairable.rs [hunk 3]
```diff
@@ -0,0 +1,135 @@
+//! Repairable Threshold Scheme
+//!
+//! Implements the Repairable Threshold Scheme (RTS) from https://eprint.iacr.org/2017/1155.
```
Comment (conradoplg) [inline frost-core/src/frost/keys/repairable.rs:3]:
```suggestion
//! Implements the Repairable Threshold Scheme (RTS) from [https://eprint.iacr.org/2017/1155].
```

File: frost-core/src/frost/keys/repairable.rs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+/// `helper_i` retains 1 `delta_j`
+
+/// Step 3 of RTS.
+///
+/// Generates the `sigma` values from all `deltas` received from `helpers`
+/// to help `participant` recover their share.
+/// `sigma` is the sum of all received `delta` and the `delta_i` generated for `helper_i`.
+///
+/// Returns a scalar
+pub fn repair_share_step_3<C: Ciphersuite>(deltas_j: &[Scalar<C>]) -> Scalar<C> {
```
Comment (conradoplg) [inline frost-core/src/frost/keys/repairable.rs:100]: I think that naming step1 / step3 / step5 may be a bit confusing. In the DKG I used step1 / step2 / step3 and documented that between each step there is a communication round. What do you think is best?

**Review by conradoplg — CHANGES_REQUESTED** [review]
Thanks for the fixes! There is one last part that I'm not sure about

**Comment by conradoplg** [comment]
@natalieesk I left an additional comment. Could you please also solve the `.gitignore` conflict, and update the branch to make it pass CI? (`git pull origin main`)

File: frost-core/src/frost/keys/repairable.rs (hunk 4 above)
Comment (dconnolly) [reply frost-core/src/frost/keys/repairable.rs:100]: agreed

**Comment by conradoplg** [comment]
@natalieesk I think now it's just missing syncing with `main` (`git fetch && git merge origin/main`) and solving the build issues (i.e. fix the test so that it works with the new `keygen_with_dealer` which now returns a HashMap)

File: frost-core/src/frost/keys/repairable.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (conradoplg) [reply frost-core/src/frost/keys/repairable.rs]: Created https://github.com/ZcashFoundation/frost/issues/320 to track this

File: frost-core/src/frost/keys/repairable.rs [hunk 5]
```diff
@@ -0,0 +1,133 @@
+//! Repairable Threshold Scheme
+//!
+//! Implements the Repairable Threshold Scheme (RTS) from [https://eprint.iacr.org/2017/1155].
```
Comment (conradoplg) [inline frost-core/src/frost/keys/repairable.rs:3]: I gave the wrong suggestion early, sorry :grimacing: 
To make Rust doc happy I'll go ahead and commit this after commenting

```suggestion
//! Implements the Repairable Threshold Scheme (RTS) from <https://eprint.iacr.org/2017/1155>.
```

**Pull request merged**